kerastuner/tuners/ultraband.py

```diff
@@ -12,18 +12,210 @@
+            self.queue.put(index)
+
+    def _select_candidates(self):
+        for index in sorted(
+                list(range(len(self._candidates))),
+                key=lambda i: self._candidate_score[i],
+                reverse=True,
+        )[:self._model_sequence[self._bracket_index]]:
+            self.queue.put(index)
+
+    @classmethod
+    def load(cls, filename):
+        pass
+
+    def save(self):
+        pass
+
+    def _new_instance(self, space):
+        """Fill a given hyperparameter space with values.
+
+        Args:
+            space: A list of HyperParameter objects
+                to provide values for.
+
+        Returns:
+            A dictionary mapping parameter names to suggested values.
+            Note that if the Oracle is keeping tracking of a large
+            space, it may return values for more parameters
+            than what was listed in `space`.
+        """
+        self.update_space(space)
+        collisions = 0
+        while 1:
+            # Generate a set of random values.
+            values = {}
+            for p in space:
+                values[p.name] = p.random_sample(self._seed_state)
+                self._seed_state += 1
+            # Keep trying until the set of values is unique,
+            # or until we exit due to too many collisions.
+            values_hash = self._compute_values_hash(values)
+            if values_hash in self._tried_so_far:
+                collisions += 1
+                if collisions > self._max_collisions:
+                    return None
+                continue
+            self._tried_so_far.add(values_hash)
+            break
+        values['epochs'] = self._epoch_sequence[self._bracket_index]
+        return values
+
+    @staticmethod
+    def _get_num_brackets(factor, min_epochs, max_epochs):
+        """Compute the number of brackets based on the scaling factor"""
+        n = 1
+        v = min_epochs
+        while v < max_epochs:
+            v *= factor
+            n += 1
+        return n
+
+    def _get_model_sequence(self, factor, min_epochs, max_epochs):
+        sizes = []
+        size = min_epochs
+        for _ in range(self._num_brackets - 1):
+            sizes.append(int(size))
+            size *= factor
+        sizes.append(max_epochs)
+        sizes.reverse()
+        return sizes
+
+    def _get_epoch_sequence(self, factor, min_epochs, max_epochs):
+        """Compute the sequence of epochs per bracket."""
+        sizes = []
+        size = min_epochs
+        for _ in range(self._num_brackets - 1):
+            sizes.append(int(size))
+            size *= factor
+        sizes.append(max_epochs)
+
+        previous_size = 0
+        output_sizes = []
+        for size in sizes:
+            output_sizes.append(size - previous_size)
+            previous_size = size
+        return output_sizes
 
 
 class UltraBand(tuner_module.Tuner):
+    """Variation of HyperBand algorithm."""
```

Comment:
Add docstring with arguments descriptions.

Also: you are instantiating the default `UltraBandOracle()`, but presumably some of these arguments should be exposed in the Tuner signature so the user can configure them, no?

Reply:
Yes, I will add them in the args.